File: src/java/graph/FordFulkerson.java
package java.graph;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.Scanner;

public class FordFulkerson {

  private final int V;

  public FordFulkerson(int V) {
    this.V = V;
  }

  /**
   * Implements the Edmonds-Karp algorithm (a specific implementation of Ford-Fulkerson) using BFS
   * to find augmenting paths in the residual graph.
   *
   * @param graph The capacity graph.
   * @param s The source vertex.
   * @param t The sink (terminal) vertex.
   * @return The maximum possible flow from source to sink.
   */
  public int fordFulkerson(int[][] graph, int s, int t) {
    int u, v;

    // Create a residual graph and fill it with the capacities of the original graph.
    int[][] rGraph = new int[V][V];
    for (u = 0; u < V; u++) {
      System.arraycopy(graph[u], 0, rGraph[u], 0, V);
    }

    // This array is filled by BFS and stores the path.
    int[] parent = new int[V];

    int maxFlow = 0; // There is no flow initially

    // Augment the flow while there is a path from source to sink
    while (bfs(rGraph, s, t, parent)) {
      // Find the maximum flow through the path found.
      int pathFlow = Integer.MAX_VALUE;
      for (v = t; v != s; v = parent[v]) {
        u = parent[v];
        pathFlow = Math.min(pathFlow, rGraph[u][v]);
      }

      // Update residual capacities of the edges and reverse edges along the path
      for (v = t; v != s; v = parent[v]) {
        u = parent[v];
        rGraph[u][v] -= pathFlow;
        rGraph[v][u] += pathFlow;
      }

      // Add path flow to overall flow
      maxFlow += pathFlow;
    }

    return maxFlow;
  }

  /**
   * A BFS-based function to find if there is a path from source 's' to sink 't' in the residual
   * graph. Also fills parent[] to store the path.
   *
   * @param rGraph The residual graph.
   * @param s The source vertex.
   * @param t The sink vertex.
   * @param parent The array to store the augmenting path.
   * @return true if a path is found, false otherwise.
   */
  private boolean bfs(int[][] rGraph, int s, int t, int[] parent) {
    boolean[] visited = new boolean[V];
    Arrays.fill(visited, false);

    Queue<Integer> queue = new ArrayDeque<>();
    queue.add(s);
    visited[s] = true;
    parent[s] = -1;

    while (!queue.isEmpty()) {
      int u = queue.poll();

      for (int v = 0; v < V; v++) {
        if (!visited[v] && rGraph[u][v] > 0) {
          queue.add(v);
          parent[v] = u;
          visited[v] = true;
        }
      }
    }

    // If we reached sink in BFS starting from source, then return true, else false
    return (visited[t]);
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter the number of vertices: ");
    int n = scanner.nextInt();
    FordFulkerson ff = new FordFulkerson(n);
    int[][] graph = new int[n][n];

    while (true) {
      System.out.println("\n******* MENU *******");
      System.out.println("Press '1' to enter an edge.");
      System.out.println("Press '2' to calculate Max Flow.");
      System.out.println("Press '3' to exit.");
      int choice = scanner.nextInt();

      switch (choice) {
        case 1:
          System.out.print("Enter the source vertex: ");
          int u = scanner.nextInt();
          System.out.print("Enter the destination vertex: ");
          int v = scanner.nextInt();
          System.out.print("Enter the capacity: ");
          int capacity = scanner.nextInt();
          if (u < 0 || v < 0 || u >= n || v >= n || capacity < 0) {
            System.out.println("Invalid input. Please try again.");
            continue;
          }
          graph[u][v] = capacity;
          break;
        case 2:
          System.out.print("Enter the source vertex: ");
          int s = scanner.nextInt();
          System.out.print("Enter the sink vertex: ");
          int t = scanner.nextInt();
          if (s < 0 || t < 0 || s >= n || t >= n) {
            System.out.println("Invalid source or sink. Please try again.");
            continue;
          }
          System.out.println("The maximum possible flow is " + ff.fordFulkerson(graph, s, t));
          break;
        case 3:
          scanner.close();
          return;
        default:
          System.out.println("Invalid choice. Please try again.");
      }
    }
  }
}

File: src/cpp/graph/ford_fulkerson.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

/**
 * @brief Finds if a path exists from source to sink in the residual graph using
 * BFS.
 *
 * @param rGraph The residual graph.
 * @param s The source vertex.
 * @param t The sink vertex.
 * @param parent A vector to store the path.
 * @return true if a path is found, false otherwise.
 */
bool bfs(const vector<vector<int>> &rGraph, int s, int t, vector<int> &parent) {
  int V = rGraph.size();
  vector<bool> visited(V, false);
  queue<int> q;

  q.push(s);
  visited[s] = true;
  parent[s] = -1;

  while (!q.empty()) {
    int u = q.front();
    q.pop();

    for (int v = 0; v < V; v++) {
      if (!visited[v] && rGraph[u][v] > 0) {
        q.push(v);
        parent[v] = u;
        visited[v] = true;
      }
    }
  }
  // If we reached sink in BFS starting from source, then return true
  return visited[t];
}

/**
 * @brief Implements the Edmonds-Karp algorithm to find the maximum flow.
 *
 * This is an implementation of the Ford-Fulkerson method that uses BFS to
 * find augmenting paths in the residual graph.
 *
 * @param graph The capacity graph represented by an adjacency matrix.
 * @param s The source vertex.
 * @param t The sink vertex.
 * @return The maximum possible flow from source to sink.
 */
int fordFulkerson(const vector<vector<int>> &graph, int s, int t) {
  int V = graph.size();
  vector<vector<int>> rGraph = graph; // Create a residual graph
  vector<int> parent(V);
  int max_flow = 0;

  // Augment the flow while there is a path from source to sink
  while (bfs(rGraph, s, t, parent)) {
    // Find the maximum flow through the path found (bottleneck capacity)
    int path_flow = INT_MAX;
    for (int v = t; v != s; v = parent[v]) {
      int u = parent[v];
      path_flow = min(path_flow, rGraph[u][v]);
    }

    // Update residual capacities of the edges and reverse edges
    for (int v = t; v != s; v = parent[v]) {
      int u = parent[v];
      rGraph[u][v] -= path_flow;
      rGraph[v][u] += path_flow;
    }

    // Add path flow to overall flow
    max_flow += path_flow;
  }

  return max_flow;
}

int main() {
  int V, E;
  cout << "Enter the number of vertices: ";
  cin >> V;
  cout << "Enter the number of edges: ";
  cin >> E;

  vector<vector<int>> graph(V, vector<int>(V, 0));
  cout << "Enter the edges (source destination capacity):" << endl;
  for (int i = 0; i < E; ++i) {
    int u, v, capacity;
    cin >> u >> v >> capacity;
    if (u >= 0 && u < V && v >= 0 && v < V && capacity >= 0) {
      graph[u][v] = capacity;
    } else {
      cout << "Invalid edge or capacity, please try again." << endl;
      i--;
    }
  }

  int source, sink;
  cout << "Enter the source vertex: ";
  cin >> source;
  cout << "Enter the sink vertex: ";
  cin >> sink;

  if (source >= 0 && source < V && sink >= 0 && sink < V) {
    cout << "\nThe maximum possible flow is "
         << fordFulkerson(graph, source, sink) << endl;
  } else {
    cout << "Invalid source or sink vertex." << endl;
  }

  return 0;
}